src/libutil/source-accessor.cc

```diff
@@ -114,9 +114,29 @@ CanonPath SourceAccessor::resolveSymlinks(
                     if (!linksAllowed--)
                         throw Error("infinite symlink recursion in path '%s'", showPath(path));
                     auto target = readLink(res);
-                    res.pop();
-                    if (isAbsolute(target))
+                    if (isAbsolute(target)) {
+                        auto prefixLen = ownLocationForSymlinkResolution.size();
+                        /* TODO need to support windows paths without making
+                           Unix vulnerability. The problem is we don't know whether
+                           the target (which is *not* a `CanonPath`) is a
+                           Windows or Unix path. Because of remote stores to
+                           other OSes, and abstract stores (which should
+                           have host-OS-agnostic designs), we cannot simply
+                           use whether we are running on Unix or Windows for
+                           this. */
+                        if (!(hasPrefix(target, ownLocationForSymlinkResolution) && (target.size() == prefixLen || target[prefixLen] == '/'))) {
```

Comment:
Should `target` be canonicalized before doing `hasPrefix()` on it? Otherwise a legal symlink target like `/foo///bar` won't match `/foo/bar`.

Also I think you can use `isInDir()` here.

Reply:
Do we have code for a really weak canonicalization for this? We should not get rid of `..` for example, we should just collapse adjacent `/`.

I guess I could write it, but it is tempting to factor out the `CanonPath` iterator that does this.